validation.py

```diff
@@ -0,0 +1,492 @@
+                index = int(distance * bins / span)
+                if not np.isnan(ld[i]):
+                    r2[index] += ld[i]
+                    n[index] += 1
+                i += 1
+        # divide `r2` by `n`, but return zero where n is zero.
+        r2 = np.divide(r2, n, out=np.zeros_like(r2), where=n != 0)
+    else:
+        # Too few segregating sites to do anything meaningful.
+        # LD plots may be blank.
+        r2 = [float("nan") for _ in range(bins)]
+
+    a = f"{span//bins//1000}k"  # width of one bin, in kb
+    b = f"{span//8//1000}k"
+    c = f"{span//4//1000}k"
+    d = f"{span//2//1000}k"
+
+    return {f"$R^2$[<{a}]": r2[0],
+            f"$R^2$[{b}]": r2[bins//8],
+            f"$R^2$[{c}]": r2[bins//4],
+            f"$R^2$[{d}]": r2[bins//2]
+            }
+
+
+_simulation_functions = [
+    onepop_constantN_msprime1,
+    onepop_constantN_slim1,
+    onepop_constantN_slim2,
+    onepop_constantN_slim3,
+    twopop_no_migration_msprime1,
+    twopop_no_migration_slim1,
+    twopop_no_migration_slim2,
+    twopop_asymmetric_migration_msprime1,
+    twopop_asymmetric_migration_slim1,
+    twopop_asymmetric_migration_slim2,
+]
+
+_stats_functions = [
+    ts_properties,
+    tmrca,
+    pooled_pop_stats,
+    pairwise_pop_stats,
+    linkage_disequilibrium,
+]
+
+_default_comparisons = [
+    (onepop_constantN_msprime1, onepop_constantN_slim1),
+    (onepop_constantN_msprime1, onepop_constantN_slim2),
+    (onepop_constantN_msprime1, onepop_constantN_slim3),
+    (twopop_no_migration_msprime1, twopop_no_migration_slim1),
+    (twopop_no_migration_msprime1, twopop_no_migration_slim2),
+    (twopop_asymmetric_migration_msprime1, twopop_asymmetric_migration_slim1),
+    (twopop_asymmetric_migration_msprime1, twopop_asymmetric_migration_slim2),
+]
+
+stats_functions = {f.__name__: f for f in _stats_functions}
+simulation_functions = {f.__name__: f for f in _simulation_functions}
+default_comparisons = [(t[0].__name__, t[1].__name__)
+                       for t in _default_comparisons]
+
+
+def do_simulations(rng, output_folder, num_replicates, num_procs, sim_keys):
+    files = dict()
+    times = dict()
+    with multiprocessing.Pool(num_procs) as pool:
+        for key in sim_keys:
+            out_dir = os.path.join(output_folder, "trees", key)
+            os.makedirs(out_dir, exist_ok=True)
+            func = functools.partial(simulation_functions[key], out_dir)
+            seeds = (rng.randrange(1, 2**32) for _ in range(num_replicates))
+            res = list(pool.imap_unordered(func, seeds))
+            files[key], times[key] = zip(*res)
+            # dump timing info to a file
+            np.savetxt(os.path.join(out_dir, "times.txt"), times[key])
+    return files, times
+
+
+def find_simulations(output_folder, sim_keys):
+    files = dict()
+    times = dict()
+    for key in sim_keys:
+        out_dir = os.path.join(output_folder, "trees", key)
```

Comment:
Pathlib would make this (and the previous) function a bit shorter and easier to read.